nexus/src/app/device_auth.rs

```diff
@@ -123,20 +123,68 @@ impl super::Nexus {
+        //      would produce an expiration later than the current token's.
+        //
+        // If the user does not request a specific TTL, we do not error out.
+        // We calculate the token TTL as min(silo max TTL, current token TTL
+        // if present). Token confirm requests authenticated with a console
+        // session can get device tokens with TTLs up to the silo max.
+
+        let time_expires = if let Some(requested_ttl) = requested_ttl {
+            // If the user requested a TTL, validate it against the silo max
+            // TTL as well as the expiration time of the token being used (if a
+            // token is being used)
+
+            // Validate the requested TTL against the silo's max TTL
+            if let Some(max) = silo_max_ttl {
+                if requested_ttl > max.0.into() {
+                    return Err(Error::invalid_request(&format!(
+                        "Requested TTL {} seconds exceeds maximum allowed \
+                     TTL for this silo of {} seconds",
+                        requested_ttl, max
+                    )));
+                }
+            };
+
+            let requested_exp =
+                Utc::now() + Duration::seconds(requested_ttl.0.into());
+
+            // If currently authenticated via token, error if requested exceeds it
+            if let Some(auth_exp) = opctx.authn.device_token_expiration() {
+                if requested_exp > auth_exp {
+                    return Err(Error::invalid_request(
+                        "Requested token TTL would exceed the expiration time \
+                         of the token being used to authenticate the confirm \
+                         request. To get the full requested TTL, confirm \
+                         this token using a web console session. Alternatively, \
+                         omit requested TTL to get a token with the longest \
+                         allowed lifetime, determined by the lesser of the silo \
+                         max and the current token's expiration time.",
+                    ));
+                }
             }
-        }
 
-        let time_expires = requested_ttl
-            .or(silo_max_ttl)
-            .map(|ttl| Utc::now() + Duration::seconds(ttl.0.into()));
+            Some(requested_exp)
+        } else {
+            // No explicit TTL requested. Rather than erroring out if silo max
+            // exceeds expiration time of current token, just clamp.
+            let silo_max_exp = silo_max_ttl
+                .map(|ttl| Utc::now() + Duration::seconds(ttl.0.into()));
+            match (silo_max_exp, opctx.authn.device_token_expiration()) {
+                (Some(silo_exp), Some(token_exp)) => {
+                    Some(silo_exp.min(token_exp))
+                }
+                (Some(silo_exp), None) => Some(silo_exp),
+                (None, Some(token_exp)) => Some(token_exp),
+                (None, None) => None,
+            }
```

Comment:
I'm pretty sure this is just:
```suggestion
            silo_max_exp.min(opctx.authn.device_token_expiration()))
```
given that there's [an `impl<T: Ord> Ord for Option<T>`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#impl-Ord-for-Option%3CT%3E); this type of manual `match` shouldn't be necessary..

Reply:
I was excited when I found that, but unfortunately the result of `a.min(b)` is always `None` when either `a` or `b` is `None` because `None` is less than any `Some(_)`.

```
a            b            match        a.min(b)
------------------------------------------------------------
Some(5)      Some(3)      Some(3)      Some(3)
Some(3)      Some(5)      Some(3)      Some(3)
Some(5)      None         Some(5)      None
None         Some(3)      Some(3)      None
None         None         None         None
```

<details>
<summary>test code</summary>

```rs
fn main() {
    let test_cases = vec![
        (Some(5), Some(3)),
        (Some(3), Some(5)),
        (Some(5), None),
        (None, Some(3)),
        (None, None),
    ];

    println!("{:<12} {:<12} {:<12} {:<12}", "a", "b", "match", "a.min(b)",);
    println!("{}", "-".repeat(60));

    for (a, b) in test_cases {
        let match_result = match (a, b) {
            (Some(a), Some(b)) => Some(a.min(b)),
            (Some(a), None) => Some(a),
            (None, Some(b)) => Some(b),
            (None, None) => None,
        };

        let ord_result = a.min(b);

        let format_opt = |opt: Option<i32>| match opt {
            Some(v) => format!("Some({})", v),
            None => "None".to_string(),
        };

        println!(
            "{:<12} {:<12} {:<12} {:<12}",
            format_opt(a),
            format_opt(b),
            format_opt(match_result),
            format_opt(ord_result),
        );
    }
}
```

</details>